lib/modules/platform/azure/pr-cache.ts

```diff
@@ -0,0 +1,152 @@
+import type { GitPullRequest } from 'azure-devops-node-api/interfaces/GitInterfaces';
+import { dequal } from 'dequal';
+import { DateTime } from 'luxon';
+import { logger } from '../../../logger';
+import * as memCache from '../../../util/cache/memory';
+import { getCache } from '../../../util/cache/repository';
+import type * as AzureApi from './azure-got-wrapper';
+import type { AzurePr, AzurePrCacheData } from './types';
+import { getRenovatePRFormat } from './util';
+
+export class AzurePrCache {
+  private items: AzurePr[] = [];
+  private cache: AzurePrCacheData;
+
+  private constructor(
+    private repoId: string,
+    private project: string,
+  ) {
+    const repoCache = getCache();
+    repoCache.platform ??= {};
+    repoCache.platform.azure ??= {};
+
+    let pullRequestCache = repoCache.platform.azure
+      .pullRequestsCache as AzurePrCacheData;
+    if (!pullRequestCache) {
+      logger.debug('Initializing new PR cache at repository cache');
+      pullRequestCache = {
+        items: {},
+        updated_at: null,
+      };
+    }
+    repoCache.platform.azure.pullRequestsCache = pullRequestCache;
+    this.cache = pullRequestCache;
+    this.updateItems();
+  }
+
+  private static async init(
+    repo: string,
+    project: string,
+    azureApi: any,
+  ): Promise<AzurePrCache> {
+    const res = new AzurePrCache(repo, project);
+    const isSynced = memCache.get<true | undefined>('azure-pr-cache-synced');
+
+    if (!isSynced) {
+      await res.sync(azureApi);
+      memCache.set('azure-pr-cache-synced', true);
+    }
+
+    return res;
+  }
+
+  private getPrs(): AzurePr[] {
+    return this.items;
+  }
+
+  static async getPrs(
+    repo: string,
+    project: string,
+    azureApi: any,
+  ): Promise<AzurePr[]> {
+    const prCache = await AzurePrCache.init(repo, project, azureApi);
+    return prCache.getPrs();
+  }
+
+  private setPr(pr: AzurePr): void {
+    logger.debug(`Adding PR #${pr.number} to the PR cache`);
+    this.cache.items[pr.number] = pr;
+    this.updateItems();
+  }
+
+  static async setPr(
+    repo: string,
+    project: string,
+    azureApi: any,
+    item: AzurePr,
+  ): Promise<void> {
+    const prCache = await AzurePrCache.init(repo, project, azureApi);
+    prCache.setPr(item);
+  }
+
+  private reconcile(rawItems: GitPullRequest[]): boolean {
+    const { items: oldItems } = this.cache;
+    let { updated_at } = this.cache;
+
+    let needNextPage = true;
+
+    for (const rawItem of rawItems) {
+      const id = rawItem.pullRequestId!;
+
+      const oldItem = oldItems[id];
+      const newItem = getRenovatePRFormat(rawItem);
+
+      // Using current date because azure pullrequest responses do not return updated date
```

Comment:
does it mean we query all PRs on each run? 🤔

Reply:
No. We do one fetch request and compare it with cache. If we found any one PR similar between the fetched and cached we stop fetching PRs.

Reply:
do we need the `updated_at` at all then? looks like it doesn't make sense if the pr's don't have an updated date

😞 we need to load all pr's each time because we never know which of the PR's have changed because they're always returned in pr id order decending.